File: development/NET/samples/server/ServerDLL/CS/OpcProject/MyDaAddressSpaceElement.cs
using System;
using System.Collections;
using Softing.OPCToolbox;
using Softing.OPCToolbox.Server;

namespace DllSample
{
	public class MyDaAddressSpaceElement : DaAddressSpaceElement
	{

		#region Constructors
		//------------------------------

		public MyDaAddressSpaceElement(
			string anItemID,
			string aName,
			uint anUserData,
			uint anObjectHandle,
			uint aParentHandle)
			:
			base(
			anItemID,
			aName,
			anUserData,
			anObjectHandle,
			aParentHandle)
		{
		}   //	end constructor

		public MyDaAddressSpaceElement()
		{
		}   //	end constructor

		//--
		#endregion

		#region Public Static Attributes
		//-------------------------------

		static public byte TYPE_UNKNOWN = 0x0;
		static public byte TYPE_NODEMATH = 0x80;
		static public byte TYPE_SINE = 0x04;
		static public byte TYPE_ACCEPT = 0x08;

		//--
		#endregion

		#region Private Attributes
		//-------------------------------

		private byte m_Type = TYPE_UNKNOWN;
		private Hashtable m_properties = new Hashtable();

		//--
		#endregion

		#region Public Property
		//---------------------

		public virtual byte Type
		{
			get { return m_Type; }
			set { m_Type = value; }
		}   //	Attribute Type

		//--
		#endregion

		#region Public Methods
		//---------------------

		/// <summary>
		/// Get elements property value data
		/// </summary>
		public void GetPropertyValue(DaRequest aRequest)
		{
			if (aRequest.PropertyId == 101)
			{
				aRequest.Value = new ValueQT(
					"description",
					EnumQuality.GOOD,
					DateTime.Now);

				aRequest.Result = EnumResultCode.S_OK;
			}
			else
			{
				aRequest.Result = EnumResultCode.E_NOTFOUND;
			}   //	end if ... else
		}   //	end GetPropertyValue

		public override int QueryProperties(out ArrayList aPropertyList)
		{
			if (m_properties.Count > 0)
			{
				aPropertyList = new ArrayList();
				aPropertyList.AddRange(m_properties.Values);
			}
			else
			{
				aPropertyList = null;
			}   //	end if ... else
			return (int)(EnumResultCode.S_OK);
		}   //	end QueryProperties


		public int AddProperty(DaProperty aProperty)
		{
			if (aProperty != null)
			{
				m_properties.Add(
					aProperty.Id,
					aProperty);
				return (int)EnumResultCode.S_OK;
			}   //	end if
			else
			{
				return (int)EnumResultCode.S_FALSE;
			}   //	end if...else
		}   //	end AddProperty

		//--
		#endregion

	}   //	end class MyAddressSpaceElement
}   //	end namespace

File: development/NET/samples/server/ServerDLL/VB/OpcProject/MyDaAddressSpaceElement.vb
Imports System
Imports System.Collections

Imports Softing.OPCToolbox
Imports Softing.OPCToolbox.Server

Namespace DllSample
	Public Class MyDaAddressSpaceElement
		Inherits DaAddressSpaceElement

#Region "Constructors"
		'------------------------------

		Public Sub New(ByVal anItemID As String, ByVal aName As String, ByVal anUserData As System.UInt32, ByVal anObjectHandle As UInt32, ByVal aParentHandle As UInt32)
			MyBase.New(anItemID, aName, anUserData, anObjectHandle, aParentHandle)
		End Sub

		Public Sub New()
		End Sub

		'--
#End Region

#Region "Public Static Attributes"
		'-------------------------------

		Public Shared TYPE_UNKNOWN As Byte = 0
		Public Shared TYPE_NODEMATH As Byte = 128
		Public Shared TYPE_SINE As Byte = 4
		Public Shared TYPE_ACCEPT As Byte = 8

		'--
#End Region

#Region "Private Attributes"
		'-------------------------------

		Private m_Type As Byte = TYPE_UNKNOWN
		Private m_properties As New Hashtable

		'--
#End Region

#Region "Public Property"
		'---------------------

		Public Overridable Property Type() As Byte
			Get
				Return m_Type
			End Get
			Set(ByVal Value As Byte)
				m_Type = Value
			End Set
		End Property

		'--
#End Region

#Region "Public Methods"
		'---------------------

		''' <summary>
		''' Get elements property value data
		''' </summary>
		Public Sub GetPropertyValue(ByVal aRequest As DaRequest)
			If aRequest.PropertyId = 101 Then
				aRequest.Value = New ValueQT("description", [Enum].ToObject(GetType(EnumQuality), EnumQuality.GOOD), DateTime.Now)

				aRequest.Result = EnumResultCode.S_OK
			Else
				aRequest.Result = EnumResultCode.E_NOTFOUND
			End If
		End Sub

		Public Overloads Overrides Function QueryProperties(ByRef aPropertyList As ArrayList) As Integer
			If m_properties.Count > 0 Then
				aPropertyList = New ArrayList
				aPropertyList.AddRange(m_properties.Values)
			Else
				aPropertyList = Nothing
			End If
			Return CType((EnumResultCode.S_OK), Integer)
		End Function


		Public Function AddProperty(ByVal aProperty As DaProperty) As Integer
			If Not aProperty Is Nothing Then
				m_properties.Add(aProperty.Id, aProperty)
				Return CType(EnumResultCode.S_OK, Integer)
			Else
				Return CType(EnumResultCode.S_FALSE, Integer)
			End If
		End Function

		'--
#End Region

	End Class
End Namespace

